scripts/patterns-reorg/update_internal_links.py

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3.11
+
+# This code is a Qiskit project.
+#
+# (C) Copyright IBM 2024.
+#
+# This code is licensed under the Apache License, Version 2.0. You may obtain a
+# copy of this license in the LICENSE file in the root directory of this source
+# tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this copyright
+# notice, and modified files need to carry a notice indicating that they have
+# been altered from the originals.
+
+from __future__ import annotations
+
+import re
+import json
+from pathlib import Path
+import os 
+import glob
+
+def get_redirects(json_path: str)-> dict[str, str]:
+    dir_path = os.path.dirname(os.path.realpath(__file__))
+    with open(f"{dir_path}/{json_path}") as redirects_file:
+        return json.load(redirects_file)
+
+
+def update_link(markdown: str, folder: str, link: str, redirects: dict[str, str])-> str:
+    anchor_index = link.find("#")
+
+    if link.startswith("http") or anchor_index == 0:
+        return markdown
+    
+    if anchor_index == -1:
+        link_without_anchor = link
+        anchor = ""
+    else:
+        link_without_anchor = link[:anchor_index]
+        anchor = link[anchor_index:]
+
+    link_split = link_without_anchor.split('/')
+    if link.startswith("/") or link.startswith("../"):
+        search_key = f"{link_split[-2]}/{link_split[-1]}"
+    else:
+        search_key = f"{folder}/{link_split[-1]}"
+
+    if search_key in redirects:
+        new_link = redirects[search_key]
+        if new_link == "":
+            new_link = "/guides/"
+        return markdown.replace(markdown, markdown.replace(link, f"{new_link}{anchor}"))
+    
+    return markdown
```

Comment:
```suggestion
    if search_key not in redirects:
        return

    new_link = redirects[search_key]
    if new_link == "":
        new_link = "/guides/"
    return markdown.replace(markdown, markdown.replace(link, f"{new_link}{anchor}"))
```

Reply:
Also what's up with the double markdown.replace()?

Reply:
Oh, that's a leftover, thanks! Before, I was replacing the markdown of the whole file, and without the double step, links like `build/pulse` used as a relative link in the same folder as `pulse` were indistinguishable from the word "pulse" found in the docs. With the new approach, using `re.sub`, we only replace the content in the link, so it should be okay to remove it as long as we don't have it in the link text (I'll look closely at that case).